ibtk/src/lagrangian/FEDataManager.cpp

```diff
@@ -1199,9 +1236,207 @@ FEDataManager::prolongData(const int f_data_idx,
     F_petsc_vec->restore_array();
     X_petsc_vec->restore_array();
 
-    IBTK_TIMER_STOP(t_prolong_data);
+    IBTK_TIMER_STOP(t_prolong_data_side);
     return;
-} // prolongData
+} // prolongData_side
+
+void
+FEDataManager::prolongData_cell(const int f_data_idx,
+                                NumericVector<double>& F_vec,
+                                NumericVector<double>& X_vec,
+                                const std::string& system_name,
+                                const bool is_density,
+                                const bool accumulate_on_grid,
+                                const bool close_F,
+                                const bool close_X)
+{
+    IBTK_TIMER_START(t_prolong_data_cell);
+
+    // Extract the mesh.
+    const MeshBase& mesh = d_es->get_mesh();
+    const unsigned int dim = mesh.mesh_dimension();
+    TBOX_ASSERT(dim == NDIM);
+
+    // Extract the FE systems and DOF maps, and setup the FE object.
+    System& F_system = d_es->get_system(system_name);
+    const unsigned int n_vars = F_system.n_vars();
+    TBOX_ASSERT(n_vars == 1); // specialized to cell-centered scalar valued data
+    const DofMap& F_dof_map = F_system.get_dof_map();
+    SystemDofMapCache& F_dof_map_cache = *getDofMapCache(system_name);
+    System& X_system = d_es->get_system(COORDINATES_SYSTEM_NAME);
+    const DofMap& X_dof_map = X_system.get_dof_map();
+    SystemDofMapCache& X_dof_map_cache = *getDofMapCache(COORDINATES_SYSTEM_NAME);
+    std::vector<std::vector<unsigned int> > F_dof_indices(n_vars);
+    std::vector<std::vector<unsigned int> > X_dof_indices(NDIM);
+    FEType F_fe_type = F_dof_map.variable_type(0);
+    for (unsigned i = 0; i < n_vars; ++i) TBOX_ASSERT(F_dof_map.variable_type(i) == F_fe_type);
+    FEType X_fe_type = X_dof_map.variable_type(0);
+    for (unsigned d = 0; d < NDIM; ++d) TBOX_ASSERT(X_dof_map.variable_type(d) == X_fe_type);
+    UniquePtr<FEBase> F_fe_autoptr(FEBase::build(dim, F_fe_type)), X_fe_autoptr;
```

Comment:
Please use `std::unique_ptr` everywhere (it is exactly the same as `libMesh::UniquePtr`)